api/types/registry/registry.go

```diff
@@ -117,3 +121,143 @@ type DistributionInspect struct {
+	// instance, Prefix:"docker.io/opensuse" will match a `docker pull
+	// opensuse:tumleweed`.
+	URL RegURL `json:"Prefix"`
+	// The mirrors will be selected prior to the registry during lookup of
+	// endpoints.
+	Mirrors []Mirror `json:"Mirrors,omitempty"`
+}
+
+// NewRegistry returns a Registry and interprets input as a URL.
+func NewRegistry(input string) (Registry, error) {
+	reg := Registry{}
+	err := reg.URL.Parse(input)
+	return reg, err
+}
+
+// AddMirror interprets input as a URL and adds it as a new mirror.
+func (r *Registry) AddMirror(input string) error {
+	mir, err := NewMirror(input)
+	if err != nil {
+		return err
+	}
+	r.Mirrors = append(r.Mirrors, mir)
+	return nil
+}
+
+// ContainsMirror returns true if the URL of any mirror equals input.
+func (r *Registry) ContainsMirror(input string) bool {
+	for _, m := range r.Mirrors {
+		if m.URL.String() == input {
+			return true
+		}
+	}
+	return false
+}
+
+// Mirror holds information for a given registry mirror.
+type Mirror struct {
+	// The URL of the mirror.
+	URL RegURL `json:"URL,omitempty"`
+}
+
+// NewMirror returns a Registry and interprets input as a URL.
+func NewMirror(input string) (Mirror, error) {
+	mir := Mirror{}
+	err := mir.URL.Parse(input)
+	return mir, err
+}
+
+// RegURL is a wrapper for url.URL to unmarshal it from the JSON config and to
+// make it an embedded type for its users.
+type RegURL struct {
+	// rURL is a simple url.URL.  Notice it is no pointer to avoid potential
+	// null pointer dereferences.
+	rURL url.URL
+}
+
+// UnmarshalJSON unmarshals the byte array into the RegURL pointer.
+func (r *RegURL) UnmarshalJSON(b []byte) error {
+	var input string
+	if err := json.Unmarshal(b, &input); err != nil {
+		return err
+	}
+	return r.Parse(input)
+}
+
+// MarshalJSON marshals the RegURL.
+func (r *RegURL) MarshalJSON() ([]byte, error) {
+	return json.Marshal(r.String())
+}
+
+// Parse parses input as a URL.
+func (r *RegURL) Parse(input string) error {
+	input = strings.ToLower(input)
+	uri, err := url.Parse(input)
+	if err == nil {
+		r.rURL = *uri
+	} else {
+		return err
+	}
+	// default to https if no URI scheme is specified
+	if uri.Scheme == "" {
+		// we have to parse again to update all associated data
+		return r.Parse("https://" + input)
+	}
+
+	// sanity checks
+	if uri.Scheme != "http" && uri.Scheme != "https" {
+		return fmt.Errorf("invalid url: unsupported scheme %q in %q", uri.Scheme, uri)
+	}
+	if uri.Host == "" {
+		return fmt.Errorf("invalid url: unspecified hostname in %s", uri)
+	}
+	if uri.User != nil {
+		// strip password from output
+		uri.User = url.UserPassword(uri.User.Username(), "xxxxx")
+		return fmt.Errorf("invalid url: username/password not allowed in URI %q", uri)
+	}
+
+	return nil
+}
+
+// Host returns the host:port of the URL.
+func (r *RegURL) Host() string {
+	return r.rURL.Host
+}
+
+// Prefix returns the host:port/path of the URL.
+func (r *RegURL) Prefix() string {
+	return r.rURL.Host + r.rURL.Path
```

Comment:
This looks to be mixing hostname with prefix, which is against the goal. The goal is to decouple these, so the prefix will still be something different, while the URL is the actual registry location. We shouldn't try to repack that into a new image name that has the new registry url.

Reply:
@stevvooe I believe we refer to the same thing and that the code semantically achieves what you desire. I try to give a concrete example why `func (r *RegURL) Prefix() string` returns `host:port/path`, so we have something concrete to discuss.

**Example** (verbose)
Let's assume we configure the daemon with `"Prefix": "http://registry.com:5000/foo"`, where the `http://` specifies that the registry `registry.com:5000` will be an insecure one while the prefix for it is `foo`. So if any image reference for `registry.com:5000` is prefixed with `foo`, this registry and its specified mirrors will be used.

When doing a `docker pull registry.com:5000/foo/bar:latest`, we end up in `lookupV2Endpoints(reference string)` (see https://github.com/moby/moby/pull/34319/files#diff-0b84f69edd22d3597c88124e0172a47e). Prior to this change, the argument was just the hostname, but now it's the entire reference of the image, in case of our example it is `registry.com:5000/foo/bar:latest`. Hence, when looking up a registry endpoint, whose prefix matches the reference, we need to prepend the `host:port/` part to the prefix. So although semantically the prefix would only be the `path`, we technically need to use the entire `host:port/prefix` to check if this string prefixes the `reference` string.

Does it reflect what you expect the prefix semantics to do?

The current PR adds testing to that semantics, for the Prefix semantics for endpoint lookup see https://github.com/moby/moby/pull/34319/files#diff-58141b9e3849201e1236b32f045b679b (`TestFindRegistry`).

Reply:
Yes, so "decoupling" makes it so that an image reference is not tied to the physical location of the registry.


In addition, making prefix an URL wouldn't match reality, because image references are _not_ URL's. For example;

```bash
$ curl -v "https://docker.io/library/busybox:latest"
*   Trying 52.3.153.154...
* TCP_NODELAY set
* Connected to docker.io (52.3.153.154) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: *.docker.io
* Server certificate: Amazon
* Server certificate: Amazon Root CA 1
* Server certificate: Starfield Services Root Certificate Authority - G2
> GET /library/busybox:latest HTTP/1.1
> Host: docker.io
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Content-length: 0
< Location: https://www.docker.com/library/busybox:latest
< Connection: close
< 
* Closing connection 0
```

Whereas; `docker pull docker.io/library/busybox:latest` actually fetches the manifest from a completely different URL:

```
msg="Calling POST /v1.36/images/create?fromImage=busybox&tag=latest"
msg="Trying to pull busybox from https://registry-1.docker.io v2"
msg="Pulling ref from V2 registry: busybox:latest"
msg="docker.io/library/busybox:latest resolved to a manifestList object with 8 entries; looking for a linux/amd64 match"
msg="found match for linux/amd64 with media type application/vnd.docker.distribution.manifest.v2+json, digest sha256:4cee1979ba0bf7db9fc5d28fb7b798ca69ae95a47c5fecf46327720df4ff352d"
```


### Docker pull pre-processing

First, let's focus on the default pre-processing;

#### Case 1:

```
docker pull hello-world
```

Doing the above, the pull code;

- detects that no tag is specified, so appends `:latest`
- detects a single component reference, so prepends the default `docker.io` prefix, and `library` org

This translates into:

```
docker.io/library/hello-world:latest
```

#### Case 2:


```
docker pull docker.io/hello-world:latest
```

- detects the default (docker.io) prefix
- detects a single component reference, so prepends the default `library` org


#### Case 3:


```
docker pull myregistry.example.com/hello-world:latest
```

- first component contains a period (`.`), so assumes this is a registry name
- registry is not the default (docker.io) prefix, so no further handling is done


#### Case 4:


```
docker pull myregistry:5000/hello-world:latest
```

- first component contains a colon (`:`), so assumes this is a registry name
- registry is not the default (docker.io) prefix, so no further handling is done

#### Case 5:


```
docker pull localhost/hello-world:latest
```

- `localhost` is special-cased, so treated as registry-name
- registry is not the default (docker.io) prefix, so no further handling is done


#### Case 6 (ambigious, non-FQDN):

```
docker pull myregistry/hello-world:latest
```

- first component is not `localhost`, and does not contain a period or colon (`:`), so assumes this is a path component
- the default (docker.io) prefix is used: `docker.io/myregistry/hello-world:latest`


#### Case 7 (ambigious, non-FQDN):

```
docker pull myregistry/foo/bar/hello-world:latest
```

- first component does not contain a period or colon (`:`), so assumes this is a path component
- the default (docker.io) prefix is used: `docker.io/myregistry/foo/bar/hello-world`


### Prefix matching

- Prefix matching should be done _after_ pre-processing (see above)
- Prefix matching should *not* match partial components
- If multiple matches are found, longest path is used
- A trailing slash is implied (?)
  - Prefixes should be unique, so having both `docker.io/library` and `docker.io/library/` should produce an error


Given the configuration below:

```json
[
  {
    "Prefix": "docker.io",
    "Mirrors": ["https://mirror1"],
  },
  {
    "Prefix": "docker.io/library",
    "Mirrors": ["https://mirror2"],
  },
  {
    "Prefix": "docker.io/library/alpine",
    "Mirrors": ["https://mirror3"],
  }
]
```

Pulling `docker.io/library/hello-world:latest`

- matches first two mirrors
- second mirror is used, because it has the longest match


Pulling `docker.io/library/alpine:latest`

- matches all mirrors
- third mirror is used, because it has the longest match

Pulling `docker.io/libraryfoobar/hello-world:latest`

- only matches first mirror, because no partial matches are allowed


### Multiple mirrors for prefix

If multiple mirrors are configured for a prefix, mirrors are tried in the order they are defined; for example:

```
"Mirrors": ["https://mirror1", "https://mirror2", "https://mirror3"],
```

Will try mirror1, then mirror2, then mirror3


### Insecure registries (non-TLS)


This may need additional discussion.


Given the following configuration:

```json
[
  {
    "Prefix": "docker.io",
    "Mirrors": ["https://mirror1", "http://localhost:5000"],
  }
]
```

Should we:

- Accept this configuration? (it's a user decision to mix secure and insecure connections)?
- Produce an error; cannot mix secure/insecure
- Produce an error; insecure mirror for `docker.io` is not allowed (?)

### Insecure registries (insecure TLS)

In addition to falling back to non-TLS, insecure registries also disable TLS verification (`InsecureSkipVerify`);

```
If InsecureSkipVerify is true, TLS accepts any certificate
presented by the server and any host name in that certificate.
In this mode, TLS is susceptible to man-in-the-middle attacks.
This should be used only for testing.
````


- How do we indicate a registry is "insecure", separate from it using a TLS connection?

### Addressing ambiguity (non FQDN references)

Prefix matching could be used to address the current ambiguity if the registry uses a non-FQDN (for example: `docker pull mylocalregistry/foo/bar`)


**Option1:** rename the `Mirrors` property to a more generic `URLs`, and used it both for mirrors, _and_ the canonical URL:

```json
[
  {
    "Prefix": "docker.io",
    "URLs": ["http://localhost:5000", "https://registry-1.docker.io"],
  }
]
```

- This configuration will try pulling from http://localhost:5000, then fall back to https://registry-1.docker.io
- Should https://registry-1.docker.io be implicitly added if not specified?

**Option2:** add an URL property to specify the canonical URL


```json
[
  {
    "Prefix": "docker.io",
    "URL": "https://registry-1.docker.io",
    "Mirrors": ["http://localhost:5000"],
  }
]
```

This configuration will try pulling from http://localhost:5000, then fall back to the canonical URL: https://registry-1.docker.io

If no canonical `URL` is set, should we:

- Automatically construct it from the prefix? (`docker.io` -> https://registry-1.docker.io, `mylocalregistry` -> `https://mylocalregistry`)
- Just attempt pulling from http://localhost:5000, and give up if that fails?

Reply:
Thank you for your detailed comments, @thaJeztah.

I am a little bit surprised about the discussion as all semantics of the PR have been baked into the commit message, testing etc. for some weeks. But I am really happy about the detailed comments as it feels we are converging to get the PR in a mergable state.

> Yes, so "decoupling" makes it so that an image reference is not tied to the physical location of the registry.

The example cases above refer to the official registry/namespace (docker.io, etc.), which is already special cased in the code. If we fully decouple it, the big delta to what the current PR does is that it would allow us to apply a full translation, e.g., from docker.io to another registry:
```json
[
  {
    "Prefix": "docker.io",
    "URL": "https://example.registry.com",
    "Mirrors": ["http://localhost:5000"],
  }
]
```
If we use an unofficial registry, the prefix will effectively match the URL (minus the URI scheme), or am I missing something? That's why the `Prefix` in the api is effectively a URL as all data can be stored in one place + parse-checking the prefix.

### Addressing ambiguity (non FQDN references)
I prefer **Option 2**. Option 1 would mix mirrors and the registry host in one field, which means that certain "magic" must be implied.

> Option2: add an URL property to specify the canonical URL
```json
[
  {
    "Prefix": "docker.io",
    "URL": "https://registry-1.docker.io",
    "Mirrors": ["http://localhost:5000"],
  }
]
```
I find this option to be the most intuitive and self-explanatory one. This structure would give us a lot of flexibility and combinations yielding different semantics:

- **URL**: allows us to specify that the registry is insecure
- **Prefix + URL**: prefix semantics for the specified URL
- **URL + Mirrors**: use mirrors as endpoints prior to the registry
- **Prefix + URL + Mirrors**: all of the above ^
- **Prefix + Mirrors**: I would allow this combination :+1: 

> If no canonical URL is set, should we: [...]

I suggest to use the mirrors then, as a user might use that to translate one prefix to a *set* of hosts.

### Prefix matching

> - Prefix matching should not match partial components

I am against enforcing this limitation. I imagine that not all registries or namespaces implement a proper path-based structure.

> A trailing slash is implied (?)

Either this or we strip trailing slashes (depending if we reject partial matches or not). Otherwise, it'll be tough to enforce uniqueness.

### Insecure registries (non-TLS)
> - Accept this configuration? (it's a user decision to mix secure and insecure connections)?

The current PR warns the user about mixed security levels and I think we should accept this configuration as rejecting it would limit many use cases. For instance, I have a mirror running on a local workstation that I use as a pull through cache. Using TLS isn't necessary for this use case.

### Insecure registries (insecure TLS)
> - How do we indicate a registry is "insecure", separate from it using a TLS connection?

We could either add a field to the JSON config, or we embed that into the URI scheme of each URL (e.g., `skip-verify://example.com`). I prefer the latter as I find it more elegant. Maybe there is another option?

Reply:
> If we fully decouple it, the big delta to what the current PR does is that it would allow us to apply a full translation, e.g., from docker.io to another registry

Correct. I think effectively the current `--registry-mirror` already allows this, as the current implementation enables you to:

- configuring `https://example.registry.com` or `http://localhost:5000` as a mirror (which will be tried _before_ contacting Docker Hub, or using "pull through")
- and: setting a firewall rule to block access to Docker Hub

I can see this possibly being the most controversial change though (given many discussions in the past), but based on the above, my _personal_ opinion is that this doesn't change what's already possible.

We need to make sure that that (like discussed with "mirrors") the user is informed that a custom registry is used, to prevent a user from unknowingly pulling from a different registry. (Again; my personal look at things, I'll check with other maintainers :sweat_smile:)

One thing to look at is if this approach has consequences for Docker Content Trust; I'll check with that team, and ask them to have a look.

> If we use an unofficial registry, the prefix will effectively match the URL (minus the URI scheme), or am I missing something? That's why the Prefix in the api is effectively a URL as all data can be stored in one place + parse-checking the prefix.

That's the current situation yes, but also reason for a lot of complaints, e.g.;

- I pushed my image to `internal-registry1/foo/bar:latest`, but the registry moved to `internal-registry2` - all my images are now invalid
- I want to switch between staging and production registries as part of our QA process

### Prefix matching

> I am against enforcing this limitation. I imagine that not all registries or namespaces implement a proper path-based structure.

Allowing partial matches, will lead to unwanted matches (`myregistry/foo` also matching `myregistry/foobar`). I'd prefer to keep this simple for now, as discussed earlier in this PR https://github.com/moby/moby/pull/34319#issuecomment-356169322

We can extend this functionality in future, using, e.g.:

- Wildcard/globbing `"Prefix": "myregistry.example.com/team-*"`
- Adding (e.g.) a `"PrefixMatch"` option that has a regex to be used for matching

> Either this or we strip trailing slashes (depending if we reject partial matches or not). Otherwise, it'll be tough to enforce uniqueness.

Agreed, we should normalize the prefix


### Insecure registries (non-TLS)

> The current PR warns the user about mixed security levels and I think we should accept this configuration as rejecting it would limit many use cases. For instance, I have a mirror running on a local workstation that I use as a pull through cache. Using TLS isn't necessary for this use case.

My reason for mentioning this was to prevent credentials from being leaked; I must admit, that didn't look at the current behavior in this respect, but we must prevent credentials that were stored for a secure connection to be sent in plain text if a mirror is insecure.

> We could either add a field to the JSON config, or we embed that into the URI scheme of each URL (e.g., `skip-verify://example.com`). I prefer the latter as I find it more elegant. Maybe there is another option?

I'd like to prevent inventing a new microformat (or "made up" scheme). @stevvooe's original suggestion was to make the mirrors a struct (`{"URL": "https://mirror1"}`), which would allow adding additional settings per-mirror, but a single configuration for all mirrors could work as well.

In general, we should discuss if we really _want_ an option to disable verification; perhaps "secure" (TLS), and "insecure" (non-TLS) as only options would suffice for the new implementation? (We can keep an internal option for mirrors set through the old flag/configuration)